src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs

```diff
@@ -191,14 +197,74 @@ public DefineFunctionsResult DefineFunctions(string script, bool numberIsFloat =
             var udfDefinitions = result.UDFs.Select(udf => new UDFDefinition(
                 udf.Ident.ToString(),
                 new ParseResult(udf.Body, errors, result.HasError, comments, null, null, script),
-                udf.ReturnType.GetFormulaType(),
+                GetFormulaTypeFromName(udf.ReturnType._value),
                 udf.IsImperative,
                 udf.NumberIsFloat,
-                udf.Args.Select(arg => new NamedFormulaType(arg.NameIdent.ToString(), arg.TypeIdent.GetFormulaType())).ToArray())).ToArray();
+                udf.Args.Select(arg => new NamedFormulaType(arg.NameIdent.ToString(), GetFormulaTypeFromName(arg.TypeIdent._value))).ToArray())).ToArray();
 
             return DefineFunctions(udfDefinitions);
         }
 
+        [Obsolete("preview")]
+        public void DefineType(string script, ParserOptions parserOptions)
+        {
+            var parsedNamedFormulasAndUDFs = UserDefinitions.Parse(script, parserOptions);
+            var definedTypes = parsedNamedFormulasAndUDFs.DefinedTypes.ToList();
+
+            foreach (var defType in definedTypes)
+            {
+                var name = defType.Ident.Name.Value;
+                var res = DTypeFromTexlNode(defType.Type.TypeRoot) ?? throw new Exception("Failed defining type");
+                _definedTypeSymbolTable.RegisterType(name, new KnownRecordType(res));
+            }
+        }
+
+        internal DType DTypeFromTexlNode(TexlNode node)
+        {
+            if (node is RecordNode recordNode)
+            {
+                return DTypeFromRecordNode(recordNode);
+            }
+            else if (node is FirstNameNode nameNode)
+            {
+                return GetTypeFromName(nameNode.Ident.Name.Value);
+            }
+
+            return null;
+        }
+
+        internal DType DTypeFromRecordNode(RecordNode recordNode)
+        {
+            var list = new List<TypedName>();
+            foreach (var (node, ident) in recordNode.ChildNodes.Zip(recordNode.Ids, (a, b) => (a, b)))

```

Comment:
This should probably be implemented as a visitor over TexlNode, but this works for now.